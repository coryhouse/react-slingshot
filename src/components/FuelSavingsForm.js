import React, {PropTypes} from 'react';
import FuelSavingsResults from './FuelSavingsResults';
import FuelSavingsTextInput from './FuelSavingsTextInput';

class FuelSavingsForm extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.save = this.save.bind(this);
    this.onTimeframeChange = this.onTimeframeChange.bind(this);
    this.fuelSavingsKeypress = this.fuelSavingsKeypress.bind(this);
  }

  onTimeframeChange(e) {
    this.props.calculateFuelSavings(this.props.fuelSavings, 'milesDrivenTimeframe', e.target.value);
  }

  fuelSavingsKeypress(name, value) {
    this.props.calculateFuelSavings(this.props.fuelSavings, name, value);
  }

  save() {
    this.props.saveFuelSavings(this.props.fuelSavings);
  }

  render() {
    const {fuelSavings} = this.props;

    return (
      <div>
        <h2>Fuel Savings Analysis</h2>
        <table>
          <tbody>
          <tr>
            <td><label htmlFor="newMpg">New Vehicle MPG</label></td>
            <td><FuelSavingsTextInput onChange={this.fuelSavingsKeypress} name="newMpg" value={fuelSavings.newMpg}/>
            </td>
          </tr>
          <tr>
            <td><label htmlFor="tradeMpg">Trade-in MPG</label></td>
            <td><FuelSavingsTextInput onChange={this.fuelSavingsKeypress} name="tradeMpg" value={fuelSavings.tradeMpg}/>
            </td>
          </tr>
          <tr>
            <td><label htmlFor="newPpg">New Vehicle price per gallon</label></td>
            <td><FuelSavingsTextInput onChange={this.fuelSavingsKeypress} name="newPpg" value={fuelSavings.newPpg}/>
            </td>
          </tr>
          <tr>
            <td><label htmlFor="tradePpg">Trade-in price per gallon</label></td>
            <td><FuelSavingsTextInput onChange={this.fuelSavingsKeypress} name="tradePpg" value={fuelSavings.tradePpg}/>
            </td>
          </tr>
          <tr>
            <td><label htmlFor="milesDriven">Miles Driven</label></td>
            <td>
              <FuelSavingsTextInput
                onChange={this.fuelSavingsKeypress}
                name="milesDriven"
                value={fuelSavings.milesDriven}/>
              miles per
              <select
                name="milesDrivenTimeframe"
                onChange={this.onTimeframeChange}
                value={fuelSavings.milesDrivenTimeframe}>
                <option value="week">Week</option>
                <option value="month">Month</option>
                <option value="year">Year</option>
              </select>
            </td>
          </tr>
          <tr>
            <td><label>Date Modified</label></td>
            <td>{fuelSavings.dateModified}</td>
          </tr>
          </tbody>
        </table>

        <hr/>

        {fuelSavings.necessaryDataIsProvidedToCalculateSavings && <FuelSavingsResults savings={fuelSavings.savings}/>}
        <input type="submit" value="Save" onClick={this.save}/>
      </div>
    );
  }
}

FuelSavingsForm.propTypes = {
  saveFuelSavings: PropTypes.func.isRequired,
  calculateFuelSavings: PropTypes.func.isRequired,
  fuelSavings: PropTypes.object.isRequired
};

export default FuelSavingsForm;
