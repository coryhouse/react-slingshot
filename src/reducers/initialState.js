export default {
  fuelSavings: {
    newMpg: '',
    tradeMpg: '',
    newPpg: '',
    tradePpg: '',
    milesDriven: '',
    milesDrivenTimeframe: 'week',
    displayResults: false,
    dateModified: null,
    necessaryDataIsProvidedToCalculateSavings: false,
    savings: {
      monthly: 0,
      annual: 0,
      threeYear: 0
    }
  }
};
