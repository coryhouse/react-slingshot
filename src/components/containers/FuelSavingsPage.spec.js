import React from "react";
import { mount, shallow } from "enzyme";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import ConnectedFuelSavingsPage, { FuelSavingsPage } from "./FuelSavingsPage";
import FuelSavingsForm from "../FuelSavingsForm";
import initialState from "../../reducers/initialState";

describe("<FuelSavingsPage />", () => {
  const actions = {
    saveFuelSavings: jest.fn(),
    calculateFuelSavings: jest.fn()
  };

  it("should contain <FuelSavingsForm />", () => {
    const wrapper = shallow(
      <FuelSavingsPage
        actions={actions}
        fuelSavings={initialState.fuelSavings}
      />
    );

    expect(wrapper.find(FuelSavingsForm).length).toEqual(1);
  });

  it("calls saveFuelSavings upon clicking save", () => {
    const wrapper = mount(
      <FuelSavingsPage
        actions={actions}
        fuelSavings={initialState.fuelSavings}
      />
    );

    const save = wrapper.find('input[type="submit"]');
    save.simulate("click");

    expect(actions.saveFuelSavings).toHaveBeenCalledWith(
      initialState.fuelSavings
    );
  });

  it("calls calculateFuelSavings upon changing a field", () => {
    const wrapper = mount(
      <FuelSavingsPage
        actions={actions}
        fuelSavings={initialState.fuelSavings}
      />
    );
    const name = "newMpg";
    const value = 10;

    const input = wrapper.find('input[name="newMpg"]');
    input.simulate("change", { target: { name, value } });

    expect(actions.calculateFuelSavings).toHaveBeenCalledWith(
      initialState.fuelSavings,
      name,
      value
    );
  });

  it("should match snapshot", () => {
    const store = configureMockStore()(initialState);
    const component = create(
      <Provider store={store}>
        <ConnectedFuelSavingsPage />
      </Provider>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
