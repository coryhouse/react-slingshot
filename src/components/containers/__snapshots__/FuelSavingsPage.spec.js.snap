// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FuelSavingsPage /> should match snapshot 1`] = `
<div>
  <h2>
    Fuel Savings Analysis
  </h2>
  <table>
    <tbody>
      <tr>
        <td>
          <label
            htmlFor="newMpg"
          >
            New Vehicle MPG
          </label>
        </td>
        <td>
          <input
            className="small"
            name="newMpg"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <label
            htmlFor="tradeMpg"
          >
            Trade-in MPG
          </label>
        </td>
        <td>
          <input
            className="small"
            name="tradeMpg"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <label
            htmlFor="newPpg"
          >
            New Vehicle price per gallon
          </label>
        </td>
        <td>
          <input
            className="small"
            name="newPpg"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <label
            htmlFor="tradePpg"
          >
            Trade-in price per gallon
          </label>
        </td>
        <td>
          <input
            className="small"
            name="tradePpg"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value=""
          />
        </td>
      </tr>
      <tr>
        <td>
          <label
            htmlFor="milesDriven"
          >
            Miles Driven
          </label>
        </td>
        <td>
          <input
            className="small"
            name="milesDriven"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value=""
          />
          miles per
          <select
            name="milesDrivenTimeframe"
            onChange={[Function]}
            value="week"
          >
            <option
              value="week"
            >
              Week
            </option>
            <option
              value="month"
            >
              Month
            </option>
            <option
              value="year"
            >
              Year
            </option>
          </select>
        </td>
      </tr>
      <tr>
        <td>
          <label>
            Date Modified
          </label>
        </td>
        <td />
      </tr>
    </tbody>
  </table>
  <hr />
  <input
    onClick={[Function]}
    type="submit"
    value="Save"
  />
</div>
`;
